File: game/src/main/java/com/bship/games/endpoints/BoardsController.java
package com.bship.games.endpoints;

import com.bship.games.domains.Board;
import com.bship.games.domains.PieceList;
import com.bship.games.exceptions.BoardValidation;
import com.bship.games.services.BoardService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.ObjectError;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

import javax.validation.Valid;
import java.util.function.Function;
import java.util.stream.Stream;

import static java.util.Optional.of;
import static org.springframework.http.HttpStatus.OK;
import static org.springframework.http.MediaType.APPLICATION_JSON_UTF8_VALUE;
import static org.springframework.http.ResponseEntity.badRequest;

@RestController
public class BoardsController implements BadRequestHandler {

    private BoardService service;

    private Function<Exception, ObjectError> pieceError;

    @Autowired
    public BoardsController(BoardService service) {
        pieceError = error.apply("board");
        this.service = service;
    }

    @PutMapping(
            value = "/boards/{boardId}",
            produces = APPLICATION_JSON_UTF8_VALUE)
    @ResponseStatus(OK)
    public Board placeShip(@PathVariable Long boardId,
                           @RequestBody @Valid PieceList pieces) throws BoardValidation {
        return service.placePiece(boardId, pieces.getPieces());
    }

    @Override
    @ExceptionHandler({BoardValidation.class})
    public ResponseEntity processValidationError(Exception check) {
        return badRequest().body(getErrors(of(check).map(pieceError).map(Stream::of).map(objectErrors)));
    }
}

File: game/src/main/java/com/bship/games/endpoints/board/BoardsController.kt
package com.bship.games.endpoints.board

import com.bship.games.endpoints.BadRequestHandler
import com.bship.games.endpoints.board.errors.exceptions.BoardValidation
import com.bship.games.endpoints.board.errors.validations.bulk.PieceList
import com.bship.games.endpoints.cabinet.entity.Board
import org.springframework.http.HttpStatus.OK
import org.springframework.http.MediaType.APPLICATION_JSON_UTF8_VALUE
import org.springframework.http.ResponseEntity
import org.springframework.http.ResponseEntity.badRequest
import org.springframework.web.bind.annotation.*
import java.util.Optional.of
import javax.validation.Valid

@RestController
open class BoardsController(private val service: BoardService) : BadRequestHandler {

    private val pieceError = BadRequestHandler.error("board")

    @PutMapping(
            value = "/boards/{boardId}",
            produces = arrayOf(APPLICATION_JSON_UTF8_VALUE))
    @ResponseStatus(OK)
    @Throws(BoardValidation::class)
    open fun placeShip(@PathVariable boardId: Long,
                       @RequestBody @Valid pieces: PieceList): Board {
        return service.placePiece(boardId, pieces.pieces)
    }

    @ExceptionHandler(BoardValidation::class)
    override fun processValidationError(check: Exception): ResponseEntity<*> {
        return badRequest().body(getErrors(
                of(check).map(pieceError)
                        .map { listOf(it) }
                        .map { BadRequestHandler.objectErrors }
        ))
    }
}
